### PR TITLE
Various changes and additions

### PR DESCRIPTION
**Pull request by SimonSapin**
This PR contains various changes and additions. I would have preferred to find a meaningful split of them into multiple PRs that I could send independently, but I expect this would have caused merge conflicts so here is everything a single linear branch on top of https://github.com/lion328/qemu-fw-cfg-rs/pull/1. Please see individual commit messages.

At a high-level, I’m trying to use QEMU’s `ramfb` device from RISC-V guest code, which for involves memory-mapped fw_cfg instead of port I/O, writing to a fw_cfg file instead of just reading, and DMA access (required for writing) instead of just the data register. RISC-V support is added to the test harness and CI in order to have coverage for the memory-mapped mode. Also as drive-by changes: supporting stable Rust, adding a file iterator.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -75,6 +76,18 @@ impl FwCfg {
...
     }
 
+    /// Build `FwCfg` for the device memory-mapped at the give base pointer.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must point to a valid fw_cfg device.
+    ///
+    /// Only one `FwCfg` value may exist at the same time for that pointer.
+    pub unsafe fn new_memory_mapped(base_ptr: *mut ()) -> Result<FwCfg, FwCfgError> {
```
Comment (lion328) [inline src/lib.rs:115]: Is this also works on x86/x86_64?

**Comment by SimonSapin** [comment]
For the sake of smaller PRs: https://github.com/lion328/qemu-fw-cfg-rs/pull/3 is the next batch

**Comment by SimonSapin** [comment]
Updated to use https://github.com/rust-embedded/riscv-rt/pull/95 instead of hard-coding the QEMU device tree address.

**Comment by SimonSapin** [comment]
Friendly ping :)

https://github.com/lion328/qemu-fw-cfg-rs/pull/3 is the next part of this

**Comment by lion328** [comment]
> Friendly ping :)
> 
> #3 is the next part of this

Sorry for a late reply, I've been quite busy lately. #3 is merged now.

File: src/lib.rs [hunk 2]
```diff
@@ -189,16 +231,41 @@ impl FwCfg {
         buf
     }
 
+    pub fn write_to_file(&mut self, file: &FwCfgFile, data: &[u8]) -> Result<(), FwCfgWriteError> {
```
Comment (lion328) [inline src/lib.rs:239]: Need a documentation here.

File: src/lib.rs [hunk 3]
```diff
@@ -301,4 +368,54 @@ impl MemoryMappedDevice {
...
+            register_low.write_volatile(address_low.to_be());
+        }
+    }
+}
+
+#[derive(Debug)]
+// NOTE: The memory layout of this struct must match this exactly:
+// https://gitlab.com/qemu-project/qemu/-/blob/v7.0.0/docs/specs/fw_cfg.txt#L177-181
+#[repr(C)]
+struct FWCfgDmaAccess {
```
Comment (lion328) [inline src/lib.rs]: Should be `FwCfgDmaAccess`, to keep things consistent.

**Comment by SimonSapin** [comment]
Sorry I forgot about this! :sweat_smile: 

Again if you prefer single-purpose PRs, https://github.com/lion328/qemu-fw-cfg-rs/pull/4 is the next batch of commits from this one.

File: src/lib.rs (hunk 3 above)
Comment (SimonSapin) [reply src/lib.rs]: Done!

File: src/lib.rs (hunk 1 above)
Comment (SimonSapin) [reply src/lib.rs:115]: `qemu-system-i386` and `qemu-system-x86_64` don’t expose a memory-mapped fw-cfg device, so there is no valid address that could be passed to this constructor. But if you somehow had an x86/x86_64 with such a device, this should work.

**Comment by lion328** [comment]
Thanks. I will go ahead and merge this larger one.

**Pull request merged**